ESQL: Enhanced `DATE_TRUNC` with arbitrary intervals

kanoshiou opened the pull request:
Originally, `DATE_TRUNC` only supported 1-month and 3-month intervals for months, and 1-year interval for years, while arbitrary intervals were supported for weeks and days. This PR adds support for `DATE_TRUNC` with arbitrary month and year intervals. 

Closes #120094

kanoshiou committed:
Add `multiplier` parameter to `roundFloor` method in `DateTimeUnit`

kanoshiou committed:
Update changelog

kanoshiou committed:
Add tests for `DateUtils`

kanoshiou committed:
Add tests for `DateTrunc`

kanoshiou committed:
Merge branch 'main' into arbitrary-intervals-of-month-year-buckets

kanoshiou commented:
Although it is possible to specify an arbitrary interval of months or years in this PR, wouldn't it be clearer to allow users to specify a reference time from which to start calculating the interval? I'm not sure.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

bpintea reviewed (commented):
I think it looks good. Left some minor comments.

* Let's rename the PR and its description to reflect the impact on `DATE_TRUNC`. `BUCKET` is a side effect (despite the issue being about it).

* Could we please add more CSV test with the new intervals, which should make it easy to check, in the lines of: `ROW x = ["1970-04-10", "1970-04-11", "1970-04-12"]::DATETIME | MV_EXPAND x | EVAL y = DATE_TRUNC(2 days, x)` and/or similar with `...| STATS COUNT(*) BY BUCKET(...)` for the newly available intervals.

CC @not-napoleon this updates some utils used by the histogram aggs, should you want to have a look.

bpintea committed:
Merge branch 'main' into arbitrary-intervals-of-month-year-buckets

bpintea commented:
@elasticsearchmachine test this please

kanoshiou committed:
Apply suggestions from code review

Co-authored-by: Bogdan Pintea <pintea@mailbox.org>

kanoshiou committed:
Add `override` annotation

kanoshiou committed:
Update testcases

kanoshiou committed:
Merge branch 'main' into arbitrary-intervals-of-month-year-buckets

kanoshiou committed:
Checkstyle

kanoshiou committed:
Add more tests

kanoshiou committed:
Merge branch 'main' into arbitrary-intervals-of-month-year-buckets

kanoshiou commented:
Thank you for reviewing, @bpintea! All comments have been addressed, and I've added some additional CSV tests.

kanoshiou committed:
reformat code

bpintea commented:
buildkite test this

kanoshiou committed:
Merge branch 'main' into arbitrary-intervals-of-month-year-buckets

kanoshiou committed:
Resolve failing tests

kanoshiou commented:
All failed tests have been resolved @bpintea .

bpintea commented:
buildkite test this

kanoshiou committed:
fix bwc

kanoshiou committed:
Merge branch 'main' into arbitrary-intervals-of-month-year-buckets

kanoshiou commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java:
```diff
@@ -547,6 +547,11 @@ public enum Cap {
          */
         BUCKET_INCLUSIVE_UPPER_BOUND,
 
+        /**
+         * Enhanced DATE_TRUNC with arbitrary month and year intervals. (#120302)
+         */
+        DATE_TRUNC_WITH_ARBITRARY_INTERVALS,
+
```
Learned from another PR, I think this should be the right way to fix those failed bwc tests. I've also run those tests on my machine, and they passed without any issues.

bpintea commented:
buildkite test this

bpintea commented:
@kanoshiou thanks for adding more tests. 
It also occurred to me, that we'll need to document how these intervals are considered, since a `DATE_TRUNC(10 months, "1969-05-01"::DATETIME)` returning `1969-03-01` might not be an obvious result (and it's maybe even more puzzling with weeks, though that's nothing new). We'll need to at least mention that the reference is the epoch, not the start of the year.

Actually, DATE_TRUNC's docs could get a refresh, to also reference BUCKET after present examples and maybe also add some examples on how to add offsets, to get alignment to years (which I think will be a common case), but that can be a follow-up.

kanoshiou committed:
Merge branch 'main' into arbitrary-intervals-of-month-year-buckets

kanoshiou committed:
Update docs

kanoshiou committed:
Merge branch 'main' into arbitrary-intervals-of-month-year-buckets

kanoshiou committed:
Update changelog

kanoshiou committed:
Merge branch 'refs/heads/main' into arbitrary-intervals-of-month-year-buckets

# Conflicts:
#	docs/reference/esql/functions/description/date_trunc.asciidoc
#	docs/reference/esql/functions/examples/bucket.asciidoc

bpintea commented:
buildkite test this

kanoshiou committed:
Merge branch 'refs/heads/main' into arbitrary-intervals-of-month-year-buckets

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Bucket.java

kanoshiou committed:
Update docs

kanoshiou commented:
Documentation conflicts have been resolved, and the branch is ready for testing again. @bpintea, could you please take another look?

bpintea commented on docs/reference/query-languages/esql/_snippets/functions/description/date_trunc.md:
```diff
@@ -2,5 +2,6 @@
 
-Rounds down a date to the closest interval.
+Rounds down a date to the closest interval since epoch, which starts
+at `0001-01-01T00:00:00Z`.
```
I believe The Epoch is the unix one in this case too, 1970. This shows in the tests too:
```
| EVAL y = DATE_TRUNC(10 months, x)
;

x:date                      | y:date
1969-05-01T00:00:00.000Z    | 1969-03-01T00:00:00.000Z
```

kanoshiou replied:
> I believe The Epoch is the unix one in this case too, 1970. 

I think this may be coincidental. 

But I discovered some implementation errors in those `roundInterval` methods and have corrected them. Additionally, the updated methods now support time calculations for dates extending to the era before the Common Era.

bpintea replied:
> I think this may be coincidental.

You're right.

I'd only have one small ask for this clarification, which is more needed when the interval to truncate to doesn't align with the calendar (for instance 1 month or 3 months does, 10 months doesn't and then the reference point needs to be understood).

Something in the lines of `Rounds down a date to the closest interval. If the interval is not calendar aligned, the intervals are counted since the beginning of the Common Era.`, or whatever fits better.

bpintea committed:
Merge branch 'main' into arbitrary-intervals-of-month-year-buckets

kanoshiou committed:
Support for negative years

kanoshiou committed:
Merge branch 'main' into arbitrary-intervals-of-month-year-buckets

kanoshiou commented:
There are some failing tests in `DateTruncTests, and I plan to address them later.

kanoshiou committed:
Update `DateTruncTests` + docs

kanoshiou committed:
Merge branch 'main' into arbitrary-intervals-of-month-year-buckets

kanoshiou commented:
I believe this PR is now ready for review. It would be great if you could take a moment to look through it @bpintea !

bpintea committed:
Merge branch 'main' into arbitrary-intervals-of-month-year-buckets

bpintea commented:
buildkite test this

kanoshiou committed:
Update tests

bpintea committed:
Merge branch 'main' into arbitrary-intervals-of-month-year-buckets

bpintea commented:
buildkite test this

bpintea reviewed (approved):
Thanks!
It LGTM, but will ask for a review from other aggs-contributors, since the changes affect those too.

not-napoleon commented:
Hi, "other aggs-contributor" here.  I am a little concerned about performance here.  This adds a lot more math, especially in the month case, and the rounding logic often runs in very hot loops so we need to be careful about performance there.  At a minimum, I'd like us to delegate the "multiplier" == 1 case to the original methods.

@bpintea we should make sure to keep an eye on the benchmarks after merging this too.

Also, it would be nice if you added some more comments about how the date math works.  This stuff is always tricky, and any help to our future selves when we need to modify it would be nice.  Thank you!

kanoshiou committed:
Minor performance improvement

kanoshiou committed:
Comments

kanoshiou commented:
Thank you for reviewing, @bpintea @not-napoleon! 

I've implemented some minor performance improvements and added detailed comments to clarify `roundIntervalMonthOfYear` and `roundYearInterval`.

bpintea commented:
buildkite test this

bpintea committed:
Merge branch 'main' into arbitrary-intervals-of-month-year-buckets

not-napoleon reviewed (approved):
This addresses my concerns from a classic aggregations perspective.  And thank you for adding more explicit comments about the date math, future me will be very happy the next time I need to touch that code.

luigidellaquila commented:
Manual backport: https://github.com/elastic/elasticsearch/pull/129643
